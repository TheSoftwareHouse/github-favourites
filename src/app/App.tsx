import React, { Component, Fragment } from 'react';
import { Route } from 'react-router-dom';

import { Search } from './search/Search';
import { ResultsContainer } from './results/ResultsContainer';

import { ApplicationContext } from './context/ApplicationContext';

import { AppState } from './App.interface';

export class App extends Component<any, AppState> {
  state: AppState = {
    favourites: [],
  };

  handleAddNewFavourite = (name: string, language: string) => {
    this.setState({
      favourites: [
        {
          name,
          language,
        },
        ...this.state.favourites,
      ],
    });
  };

  handleRemoveFavourite = (index: number) => {
    const favourites = this.state.favourites;
    favourites.splice(index, 1);
    this.setState({ favourites });
  };

  render() {
    return (
      <ApplicationContext.Provider
        value={{
          favourites: this.state.favourites,
          onAddFavourite: this.handleAddNewFavourite,
          onRemoveFavourite: this.handleRemoveFavourite,
        }}
      >
        <Route path="/" exact component={Search} />
        <Route path="/results/" component={ResultsContainer} />
      </ApplicationContext.Provider>
    );
  }
}