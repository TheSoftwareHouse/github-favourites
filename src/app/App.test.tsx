import React from 'react';
import { MemoryRouter as Router, MemoryRouter } from 'react-router-dom';

import { render, waitForElement, fireEvent } from 'react-testing-library';

import { ThemeProviderMock } from 'test/ThemeProviderMock';

import { App } from './App';

describe('App component', () => {
  it('renders app correctly', async () => {
    const { container } = render(
      <MemoryRouter>
        <ThemeProviderMock>
          <App />
        </ThemeProviderMock>
      </MemoryRouter>
    );

    expect(container).toMatchSnapshot();

    // Enter name/description
    const inputField = container.getElementsByClassName('input__field')[0];
    fireEvent.change(inputField, { target: { value: 'New favourite' } });

    // Select language
    fireEvent.click(container.getElementsByClassName('select-field__input')[0]);
    await waitForElement(() => container.getElementsByClassName('select-field__options'));
    fireEvent.click(container.getElementsByClassName('select-field__option')[0]);

    // Button available
    expect(container.getElementsByClassName('search__submit-button button--disabled').length).toEqual(0);
    expect(container.getElementsByClassName('search__submit-button button').length).toEqual(1);

    // Add to favourites
    await fireEvent.click(container.getElementsByClassName('search__submit-button button')[0]);

    // Check length of favourites
    expect(container.getElementsByClassName('favourite-box').length).toEqual(1);

    // Delete favourite
    await fireEvent.click(container.getElementsByClassName('favourite-box__menu__delete')[0]);
    expect(container.getElementsByClassName('favourite-box').length).toEqual(0);
  });
});
