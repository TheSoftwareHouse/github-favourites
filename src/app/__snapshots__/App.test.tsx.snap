// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component renders app correctly 1`] = `
<div>
  <section
    class="layout css-1jsa027"
  >
    <header
      class="header css-p0n1vq"
    >
      <div
        class="header__content css-pmsyce"
      >
        <h1
          class="header__title css-7h8kd1"
        >
          Search favourites
        </h1>
        <img
          alt="Logo"
          class="header-logo"
          src="logo.png"
        />
      </div>
    </header>
    <div
      class="css-9uvy62"
    >
      <section
        class="layout__content css-6ln5j9"
      >
        <section
          class="search__box box css-9lwejs"
        >
          <div
            class="input css-co7rad"
          >
            <input
              class="input__field"
              id="name"
              maxlength="60"
              name="name"
              value=""
            />
            <label
              class="input__label css-jrw360"
              for="name"
            >
              Repo name or description
            </label>
            <div
              class="input__counter css-17d9fe5"
            >
              0
               / 
              60
            </div>
          </div>
          <div
            class="select-field css-1tmehsk"
          >
            <div
              class="select-field__input css-7xqnnq"
            >
              <div
                class="css-crxi2m"
                color="textLightColor"
                name="chevron-arrow-down"
              />
            </div>
            <label
              class="select-field__label css-jrw360"
              for="language"
            >
              Language
            </label>
          </div>
          <div
            class="search__submit css-1qvnd68"
          >
            <button
              class="search__submit-button button button--disabled css-aenjb1"
              disabled=""
            >
              <div
                class="icon css-uivog"
                color="primaryColor"
                name="add-square-button"
              />
              Add to list
            </button>
          </div>
        </section>
        <div
          class="search__favourite-list css-16n4kcf"
        />
      </section>
    </div>
  </section>
</div>
`;
